core/src/packet_deserializer.rs

```diff
@@ -43,35 +46,65 @@ impl PacketDeserializer {
-        let (packet_batches, sigverify_tracer_stats_option) =
-            self.receive_until(recv_timeout, capacity)?;
+        let (packet_count, packet_batches) = self.receive_until(recv_timeout, capacity)?;
         Ok(Self::deserialize_and_collect_packets(
+            packet_count,
             &packet_batches,
-            sigverify_tracer_stats_option,
         ))
     }
 
+    fn for_each_packet_batch(
+        banking_batches: &[BankingPacketBatch],
+        mut for_each: impl FnMut(ForEach),
+    ) {
+        for banking_batch in banking_batches {
+            for batch in &banking_batch.0 {
+                for_each(ForEach::Batch(batch))
+            }
+
+            if let Some(stats) = &banking_batch.1 {
+                for_each(ForEach::Stats(stats))
+            }
+        }
+    }
+
     /// Deserialize packet batches and collect them into ReceivePacketResults
     fn deserialize_and_collect_packets(
-        packet_batches: &[PacketBatch],
-        sigverify_tracer_stats_option: Option<SigverifyTracerPacketStats>,
+        packet_count: usize,
+        packet_batches: &[BankingPacketBatch],
     ) -> ReceivePacketResults {
-        let packet_count: usize = packet_batches.iter().map(|x| x.len()).sum();
         let mut passed_sigverify_count: usize = 0;
         let mut failed_sigverify_count: usize = 0;
         let mut deserialized_packets = Vec::with_capacity(packet_count);
-        for packet_batch in packet_batches {
-            let packet_indexes = Self::generate_packet_indexes(packet_batch);
+        let mut aggregated_tracer_packet_stats_option = None::<SigverifyTracerPacketStats>;
 
-            passed_sigverify_count += packet_indexes.len();
-            failed_sigverify_count += packet_batch.len().saturating_sub(packet_indexes.len());
+        Self::for_each_packet_batch(packet_batches, |packet_batch_or_stat| {
+            match packet_batch_or_stat {
+                ForEach::Batch(packet_batch) => {
+                    let packet_indexes = Self::generate_packet_indexes(packet_batch);
 
-            deserialized_packets.extend(Self::deserialize_packets(packet_batch, &packet_indexes));
-        }
+                    passed_sigverify_count += packet_indexes.len();
+                    failed_sigverify_count +=
+                        packet_batch.len().saturating_sub(packet_indexes.len());
+
+                    deserialized_packets
+                        .extend(Self::deserialize_packets(packet_batch, &packet_indexes));
+                }
+                ForEach::Stats(tracer_packet_stats) => {
+                    if let Some(aggregated_tracer_packet_stats) =
+                        &mut aggregated_tracer_packet_stats_option
+                    {
+                        aggregated_tracer_packet_stats.aggregate(tracer_packet_stats);
+                    } else {
+                        aggregated_tracer_packet_stats_option = Some(tracer_packet_stats.clone());
```

Comment:
I think the introduction of the `ForEach` also makes this clone necessary because we have a reference and not an owned packet stats. We could just take the first one if it wasn't ref?

Reply:
The unavoidable introduction of `Arc`-wrapping makes this clone necessary as i commented there: https://github.com/ryoqun/solana/commit/1ec9956cc3d12b8d2f6cfa8b96eecf023e280297#diff-5fbb1074da050dbfead620306af9b99aae42e02897183d1c0884f2ec49816627R79